### PR TITLE
text: emit make commands or makefile lines

### PR DESCRIPTION
**Pull request by imjasonh**
Signed-off-by: Jason Hall <jason@chainguard.dev>

```
$ dag text
make packages/x86_64/wolfi-baselayout-20221118-r0.apk
make packages/x86_64/wget-1.21.3-r2.apk
make packages/x86_64/xz-5.4.0-r0.apk
make packages/x86_64/pax-utils-1.3.4-r2.apk
make packages/x86_64/scanelf-1.3.4-r2.apk
...
make packages/x86_64/logger-2.38.1-r0.apk
make packages/x86_64/glibc-locale-szl-2.36-r4.apk
make packages/x86_64/glibc-locale-pap-2.36-r4.apk
make packages/x86_64/glibc-locale-ky-2.36-r4.apk
make packages/x86_64/openjdk-11-jre-doc-11.0.17.8-r5.apk
```

For example, `sh -c "$(dag text)"`

...or, if we want to keep the Makefile around, we can generate the big block of `eval(call build-package`s:

```
$ dag text --type=makefile
$(eval $(call build-package,wolfi-baselayout,20221118-0))
$(eval $(call build-package,wget,1.21.3-2))
$(eval $(call build-package,xz,5.4.0-0))
$(eval $(call build-package,pax-utils,1.3.4-2))
$(eval $(call build-package,zlib,1.2.13-1))
...
$(eval $(call build-package,melange,0.2.0-0))
$(eval $(call build-package,kubectl,1.25.4-1))
$(eval $(call build-package,mailcap,2.1.53-0))
$(eval $(call build-package,grype,0.55.0-0))
$(eval $(call build-package,cosign,1.13.1-1))
```

As you can see in this example, this doesn't ensure the list is consistently sorted, only that roots of the dep graph are printed before leafs.

This change also adds understanding of `provides:` since some packages depend on "`ninja`" but that isn't actually packaged, just `provides`ed by `samurai`. We also need to take care not to reject `go` when it's `provides`ed by another package, like `go-stage0`, for example.

### CODE REVIEW

File: pkg/graph.go [hunk 1]
```diff
@@ -168,30 +181,11 @@ func (g Graph) Config(name string) *build.Configuration {
 	return nil
 }
 
-// IsSubpackage returns a bool indicating whether the package with the given name
-// is a subpackage. If the package is an origin package, or if the package is not
-// found in the graph, IsSubpackage returns false.
-func (g Graph) IsSubpackage(name string) bool {
```
Comment (luhring) [inline pkg/graph.go:174]: Thanks for cleaning this up!

File: pkg/graph.go [hunk 2]
```diff
@@ -168,30 +181,11 @@ func (g Graph) Config(name string) *build.Configuration {
...
 // order, meaning that packages earlier in the list depend on packages later in
 // the list.
 func (g Graph) Sorted() ([]string, error) {
-	sorted, err := graph.TopologicalSort(g.Graph)
-	if err != nil {
-		return nil, err
-	}
-
-	return sorted, nil
+	return graph.TopologicalSort(g.Graph)
```
Comment (luhring) [inline pkg/graph.go:188]: :chefs-kiss:

File: pkg/commands/text.go [hunk 3]
```diff
@@ -58,33 +57,56 @@ func cmdText() *cobra.Command {
...
 	}
 	text.Flags().StringVarP(&dir, "dir", "d", ".", "directory to search for melange configs")
 	text.Flags().StringVarP(&arch, "arch", "a", "x86_64", "architecture to build for")
 	text.Flags().BoolVarP(&showDependents, "show-dependents", "D", false, "show packages that depend on these packages, instead of these packages' dependencies")
+	text.Flags().StringVarP(&t, "type", "t", string(typeTarget), "What type of text to emit")
 	return text
 }
 
-func text(g pkg.Graph, arch string, w io.Writer) error {
+type textType string
```
Comment (luhring) [inline pkg/commands/text.go:70]: nit: If we're going to create this type, perhaps we should use it when defining the consts below. Right now it looks like we're only using this type in `text()`, where values are compared to `string`s directly

File: pkg/graph.go [hunk 4]
```diff
@@ -57,22 +57,37 @@ func NewGraph(dirFS fs.FS, dirPath string) (*Graph, error) {
 			}
 
 			c := build.Configuration{}
-			err = (&c).Load(*buildContext)
-			if err != nil {
+			if err := (&c).Load(*buildContext); err != nil {
```
Comment (luhring) [inline pkg/graph.go:60]: nit: At some point we should replace this call with one to [`build.ParseConfiguration`](https://github.com/chainguard-dev/melange/blob/main/pkg/build/build.go#L700-L701)

**Review by luhring — APPROVED** [review]
Very cool!!! 🎉

File: pkg/graph.go (hunk 4 above)
Comment (luhring) [reply pkg/graph.go]: Looks good! Works on my machine w/o Docker now

**Comment by kaniini** [comment]
hmm, the problem with this output is that it includes subpackages.  while those should be considered in the DAG for dependency resolution, it needs to break down to the main packages only.

**Comment by imjasonh** [comment]
> hmm, the problem with this output is that it includes subpackages. while those should be considered in the DAG for dependency resolution, it needs to break down to the main packages only.

Oh I had handled this for `--type makefile` but not for `--type target` (the default).

Handling this now:

```
$ dag text | grep glibc
make packages/x86_64/glibc-2.36-r4.apk
```

(This used to also list all the locales subpackages, etc.)

**Comment by imjasonh** [comment]
Also spot-checked that both `dag text | wc -l` and `grep "call build-package," Makefile | wc -l` return `285` lines

**Comment by kaniini** [comment]
awesome :)

**Review by kaniini — APPROVED** [review]

**Pull request merged**